storage/innobase/xtrabackup/src/read_filt.cc

```diff
@@ -201,13 +203,41 @@ static void rf_page_tracking_init(xb_read_filt_ctxt_t *ctxt,
   ctxt->filter_batch_end = 0;
 }
 
+/* Build map of all spaces which need full scan
+@return map of spaces */
+std::map<space_id_t, bool> build_space_need_full_scan() {
+  std::map<space_id_t, bool> spaces;
+
+  /* we need full scan of mysql tablespaces to identify modified pages from last
+   * backup start lsn. This is done because we skip applying logical
+   * redos (MLOG_TABLE_DYNAMIC_META) during the incremental prepare (except
+   * the last prepare). These logical redos are converted to regular redo and
+   * flushed to pages in mysql.ibd when the server process a checkpoint. So
+   * we directly take the physical changes made to innodb_dynamic_metadata
+   * since the last backup. Hence we copy all changes to mysql.ibd since last
+   * backup start_lsn instead of last backup end_lsn.
+   * We read all pages by setting read_batch_len to the size of file */
+  spaces[dict_sys_t::s_dict_space_id] = true;
+
+  /* if DDL happened on the table after the checkpoint LSN we do full scan on
+  that table. index_load_list is poped during the first scan of redo */
+  for (auto space_lsn_pair : index_load_list) {
+    if (space_lsn_pair.second > incremental_start_checkpoint_lsn)
+      spaces[space_lsn_pair.first] = true;
+  }
+
+  return (spaces);
+}
```

Comment:
the redo processing is in background thread and page-copying in another thread. What if the redo thread didn't reach the MLOG_INDEX_LOAD record and page-tracking is processing that space?

Reply:
Can we avoid this function?
Why not read `index_load_list` at `rf_page_tracking_get_next_batch` and check LSN calculation there ?

Reply:
rf_page_tracking_get_next_batch is called multiple times for each space. Doing calculation each time could be expensive operation. Building map would be one time operation.

Reply:
@rahulmalik87 marce's point is we don't need a list and a map. You can make index_load_list as a map and add entries for all MLOG_INDEX_LOAD redo entries.

Then just use the map here.

Reply:
instead of index_load_list, I would say map_full_scan/full_scan_spaces/

Reply:
changed the name.